[quant][graphmode] Integrate prepacked workaround in QuantFusion

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26939 [quant][graphmode] Integrate prepacked workaround in QuantFusion**
* #26579 [quant][graphmode] Fold prepacked weight into module

Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

jerryzh168 committed:
[quant][graphmode] Integrate prepacked workaround in QuantFusion

Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1199,37 +1199,61 @@ def test_quant_fusion(self):
         # CHECK-NOT: aten::dequantize
         %a_dequant = aten::dequantize(%a_quant)
-        %w_quant = aten::quantize_per_tensor(%w, %w_scale, %w_zero_point, %w_dtype)
+        %packed_params = prim::GetAttr[name="_packed_params"](%self)
```
This test is over-specialized and thus too fragile and not very helpful. It just replicates the pattern almost exactly - what's the point of having it? It doesn't check if the pattern is correct. At most, it will check that subgraph rewriter works, but it has its own tests. The test should not check implementation details (like `# CHECK-NOT: prim::TupleIndex`), it should rather check that the pass is doing what it's supposed to do - in this case replace `addmm`  with `quantized::linear`. For that two CHECK/CHECK_NOT directives would be enough.

jerryzh168 replied:
OK

jerryzh168 replied:
But to test it with actual patterns we see we might need e2e test

ZolotukhinM replied:
Yes, we definitely need an e2e test.

jerryzh168 replied:
many `CHECK-NOT` is part of the pattern as well, `CHECK-NOT`s are making sure we actually removed those nodes

ZolotukhinM replied:
If you find `quantized::linear` in the output then you know that the pattern was matched. There is no reason the check other details of the pattern - or you will be just copying the pattern into the test. If the pattern is buggy, you would just copy the buggy part into the test. This test cannot check if the pattern is buggy (that's the job of e2e test), and there is no need to check if the pattern rewrite was accurate (that's covered by tests for subgraph rewriter).

ZolotukhinM reviewed (approved):
Overall looks good, but please make the test less brittle before landing.

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Integrate prepacked workaround in QuantFusion"


Summary:
Updated quant fusion patterns to work with modules with prepack
params folded into module.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636398](https://our.internmc.facebook.com/intern/diff/D17636398)

[ghstack-poisoned]
